tensorflow/contrib/io/kernels/zmq_op.cc

```diff
@@ -0,0 +1,108 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// See docs in ../ops/io_ops.cc.
+
+#include <memory>
+#include <zmq.h>
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/lib/core/errors.h"
+
+namespace tensorflow {
+
+class PollZmqOp : public AsyncOpKernel {
+ public:
+  explicit PollZmqOp(OpKernelConstruction* context) : AsyncOpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("address", &address_));
+    OP_REQUIRES_OK(context, context->GetAttr("timeout", &timeout_));
+    // Create a context and connect a REQ socket
+    context_ = zmq_ctx_new();
+    OP_REQUIRES(context, context_ != nullptr,
+                errors::Internal("Failed to initialize context."));
+    socket_ = zmq_socket(context_, ZMQ_REQ);
+    OP_REQUIRES(context, socket_ != nullptr,
+                errors::Internal("Failed to initialize socket."));
+    OP_REQUIRES(context,
+                zmq_setsockopt(socket_, ZMQ_RCVTIMEO,
+                               &timeout_, sizeof(timeout_)) == 0,
+                errors::Internal("Failed to set timeout."));
+    OP_REQUIRES(context, zmq_connect(socket_, &address_[0]) == 0,
+                errors::Internal("Failed to connect to ", address_, "."));
```

Comment:
[`zmq_connect`](http://api.zeromq.org/4-2:zmq-connect) does not usually block. From the docs:

> [F]or most transports and socket types the connection is not performed immediately but as needed by ØMQ. [...] The first exception is when using the inproc:// transport: you must call zmq_bind() before calling zmq_connect(). The second exception are ZMQ_PAIR sockets, which do not automatically reconnect to endpoints.